### PR TITLE
Add couple features to duplicate model

### PR DESCRIPTION
**Pull request by ayush-1506**
Part of #523 

The results before and after are exactly the same. I'm looking into this.

### CODE REVIEW

**Comment by ayush-1506** [comment]
Update : I'm trying to figure out a way to add the couple_data as a feature from inside the extraction pipeline.

**Comment by ayush-1506** [comment]
Ok, so that works now. However, there are a few other problems:

1. Values such as "True", "False" need to be converted to integers so that we get the resultant sparse matrix. For now, I'm making the feature extractors return `int`, but I'm not sure if this is the best way.

2. How should I handle feature extractors that don't returns "true"/"false", but a list of keywords (such as `couple_common_keywords`)? We need a way to convert these keywords to numbers. (Is using a vectorizer the best option in this case?)

@marco-c

**Comment by ayush-1506** [comment]
Maybe number of common keywords make more sense than list of the common keywords?

**Comment by ayush-1506** [comment]
Results:

Adding common keywords:
```
{'test_accuracy': {'mean': 0.9663222337896421, 'std': 0.00617975339630379}, 'test_precision': {'mean': 
0.8653523448151335, 'std': 0.06047868681910123}, 'test_recall': {'mean': 0.7488813898262717, 'std': 
0.08162211314077335}, 'report': {'targets': {'1': {'precision': 0.8493150684931506, 'recall': 0.7045454545454546, 
'specificity': 0.9839181286549707, 'f1': 0.7701863354037267, 'geo_mean': 0.8325953670233395, 'iba': 
0.6738485110979969, 'support': 88}, '0': {'precision': 0.9628040057224606, 'recall': 0.9839181286549707, 'specificity': 
0.7045454545454546, 'f1': 0.9732465654374548, 'geo_mean': 0.8325953670233395, 'iba': 0.7125815792794619, 
'support': 684}}, 'average': {'precision': 0.9498674429294823, 'recall': 0.9520725388601037, 'specificity': 
0.7363910443403217, 'f1': 0.9500998034646982, 'geo_mean': 0.8325953670233395, 'iba': 0.7081664108857199, 
'support': 772}}, 'confusion_matrix': [[62, 26], [11, 673]]}
```

common keywords + common whiteboard keyword:

```
{'test_accuracy': {'mean': 0.9661783488975558, 'std': 0.006543452994697574}, 'test_precision': {'mean': 
0.8634883097274141, 'std': 0.05722827724494728}, 'test_recall': {'mean': 0.7488813898262717, 'std': 
0.08162211314077335}, 'report': {'targets': {'1': {'precision': 0.8611111111111112, 'recall': 0.7045454545454546, 'specificity': 
0.9853801169590644, 'f1': 0.775, 'geo_mean': 0.8332137075222522, 'iba': 0.6747482740680824, 'support': 88}, '0': 
{'precision': 0.9628571428571429, 'recall': 0.9853801169590644, 'specificity': 0.7045454545454546, 'f1': 
0.9739884393063584, 'geo_mean': 0.8332137075222522, 'iba': 0.7137418907378721, 'support': 684}}, 'average': 
{'precision': 0.9512591496011185, 'recall': 0.9533678756476683, 'specificity': 0.7365576958568505, 'f1': 
0.9513058192817995, 'geo_mean': 0.8332137075222521, 'iba': 0.7092970225164452, 'support': 772}}, 'confusion_matrix': 
[[62, 26], [10, 674]]}
```

Will keep adding more results as they come.

**Comment by ayush-1506** [comment]
@marco-c I'm using the following couple features right now:

> 
> common keywords + common whiteboard keyword + common words in summary + common words in comments:
> 

Let me know more features to add here.

File: bugbug/bug_features.py [hunk 1]
```diff
@@ -463,44 +463,46 @@ def __call__(self, bug, **kwargs):
 
 class couple_common_whiteboard_keywords(couple_bug_feature):
```
Comment (marco-c) [inline bugbug/bug_features.py:464]: I'd define a new feature (`couple_number_of_common_whiteboard_keywords`) instead of changing this one

File: bugbug/bug_features.py [hunk 2]
```diff
@@ -463,44 +463,46 @@ def __call__(self, bug, **kwargs):
...
+                for keyword in whiteboard_keywords(bugs[0])
+                if keyword in whiteboard_keywords(bugs[1])
+            ]
+        )
 
 
 class is_same_product(couple_bug_feature):
     def __call__(self, bugs, **kwargs):
-        return bugs[0]["product"] == bugs[1]["product"]
+        return int(bugs[0]["product"] == bugs[1]["product"])
```
Comment (marco-c) [inline bugbug/bug_features.py:477]: Why an int here and in the other features?

File: bugbug/bug_features.py (hunk 1 above)
Comment (marco-c) [reply bugbug/bug_features.py:464]: Or maybe not, if we are not using the old in any other place

File: bugbug/model.py [hunk 3]
```diff
@@ -248,10 +248,7 @@ def get_top_features(cutoff, shap_values):
 
             class_name = self.le.inverse_transform([num])[0]
 
-            important_features["classes"][class_name] = (
-                top_item_features,
-                top_avg,
-            )
+            important_features["classes"][class_name] = (top_item_features, top_avg)
```
Comment (marco-c) [inline bugbug/model.py:251]: This change seems unrelated, could you revert it?

File: bugbug/models/duplicate.py [hunk 4]
```diff
@@ -28,6 +30,28 @@ def fit(self, X, y):
         self._le.fit(y)
 
 
+class BypassFeatures(TransformerMixin, BaseEstimator):
```
Comment (marco-c) [inline bugbug/models/duplicate.py:33]: Why was this needed?

**Comment by marco-c** [comment]
> Values such as "True", "False" need to be converted to integers so that we get the resultant sparse matrix. For now, I'm making the feature extractors return int, but I'm not sure if this is the best way.

> How should I handle feature extractors that don't return "true"/"false", but a list of keywords (such as couple_common_keywords)? We need a way to convert these keywords to numbers. (Is using a vectorizer the best option in this case?)

You can solve these problems like this: https://github.com/mozilla/bugbug/blob/07590f145d5707c0efd212a72635d1e619bfe98d/bugbug/commit_features.py#L666-L672.

File: bugbug/bug_features.py (hunk 2 above)
Comment (marco-c) [reply bugbug/bug_features.py:477]: OK, read your earlier comment and I replied in https://github.com/mozilla/bugbug/pull/889#issuecomment-670501994 :)

**Comment by marco-c** [comment]
> Adding common keywords:

What are the results before adding them?

Could you try all combinations of these new features with/without xgboost?

(Could you report just accuracy, precision and recall so they're easier to read?)

File: bugbug/models/duplicate.py (hunk 4 above)
Comment (ayush-1506) [reply bugbug/models/duplicate.py:33]: Right now our pipeline looks like:
```
self.extraction_pipeline = Pipeline(
            [
                (
                    "bug_extractor",
                    bug_features.BugExtractor(feature_extractors, cleanup_functions),
                ),
                (
                    "union",
                    ColumnTransformer(
                        [
                            ("text", self.text_vectorizer(), "text"),
                        ]
                    ),
                ),
            ]
        )
```

BugFeatures returns a dictionary {"text":... , "couple_data"..}, but the `ColumnTransformer` in the next step just takes the "text" data from the dict and ignores everything else. Hence, we end up loosing the "couple_data" data, which has all the bug couple features. I confirmed this by completely removing all bug couple data, but I still get same dimensional features as obtained by using them (implying that they weren't being used at all).

One way to circumvent this is by introducing the `BypassFeatures` transformer that appends the couple data features to the current tf-idf text vectors, thus expanding our features. Is there a better way to do this? @marco-c

File: bugbug/model.py (hunk 3 above)
Comment (ayush-1506) [reply bugbug/model.py:251]: pre-commit seems to be making this change, not sure why it is detecting this

Comment (marco-c) [reply bugbug/model.py:251]: You should be able to revert it now, it was probably a different version of black

**Comment by ayush-1506** [comment]
xgboost + all feature extractors:
```
Accuracy: f0.9310097201147786 (+/- 0.036759555908687874)
Precision: f0.8938016215308388 (+/- 0.07994774867892024)
Recall: f0.8761904761904763 (+/- 0.08575999315359989)
```

**Comment by ayush-1506** [comment]
@marco-c Updated results above.

**Comment by ayush-1506** [comment]
@marco-c Not sure what's up with black: it seems to be changing a lot of files.

**Comment by marco-c** [comment]
> @marco-c Not sure what's up with black: it seems to be changing a lot of files.

You probably need to update black, you can try to:
```
pre-commit uninstall
pre-commit clean
pre-commit gc
pre-commit install
pre-commit run --all-files
```

**Comment by marco-c** [comment]
> @marco-c Updated results above.

They are looking pretty good, I hope we are not making mistakes in the evaluation code :P

**Comment by marco-c** [comment]
The patch looks good, I'll approve it once you manage to get rid of the black crazyness :P

**Comment by ayush-1506** [comment]
@marco-c r? (Had to manually revert changes for each file, not sure what's wrong)

File: bugbug/bug_features.py [hunk 5]
```diff
@@ -522,6 +522,30 @@ def __call__(self, bugs, **kwargs):
...
+        )
+
+
+class couple_common_words_comments(couple_bug_feature):
+    def __init__(self, to_ignore=set()):
+        self.to_ignore = to_ignore
+
+    def __call__(self, bugs):
+        text1 = " ".join(
+            bugs[0]["comments"][idx]["text"] for idx in range(len(bugs[0]["comments"]))
```
Comment (marco-c) [inline bugbug/bug_features.py:541]: You could rewrite as:
```
comment["text"] for comment in bug[0]["comments"]

File: bugbug/models/duplicate.py [hunk 6]
```diff
@@ -8,9 +8,10 @@
 
 from sklearn.calibration import CalibratedClassifierCV
 from sklearn.compose import ColumnTransformer
+from sklearn.feature_extraction import DictVectorizer
 from sklearn.pipeline import Pipeline
 from sklearn.preprocessing import LabelEncoder
-from sklearn.svm import LinearSVC
+from xgboost import XGBClassifier
```
Comment (marco-c) [inline bugbug/models/duplicate.py:14]: Could you split the classifier change in a separate PR?

File: bugbug/models/duplicate.py [hunk 7]
```diff
@@ -55,16 +69,23 @@ def __init__(self, training_size=14000, lemmatization=False, cleanup_urls=True):
...
+                            ("text", self.text_vectorizer(), "text"),
+                            ("couple_data", DictVectorizer(), "couple_data"),
+                        ]
+                    ),
                 ),
             ]
         )
 
-        self.clf = LinearSVCWithLabelEncoding(LinearSVC())
+        self.clf = XGBClassifier(n_jobs=4)
```
Comment (marco-c) [inline bugbug/models/duplicate.py:88]:
```suggestion
        self.clf = XGBClassifier(n_jobs=utils.get_physical_cpu_count())
```

File: bugbug/models/duplicate.py [hunk 8]
```diff
@@ -29,15 +30,28 @@ def fit(self, X, y):
 
 
 class DuplicateModel(BugCoupleModel):
-    def __init__(self, training_size=14000, lemmatization=False, cleanup_urls=True):
+    def __init__(self, training_size=2000, lemmatization=False, cleanup_urls=True):
```
Comment (marco-c) [inline bugbug/models/duplicate.py:33]: Let's revert this back to 14000

**Comment by ayush-1506** [comment]
Made the changes. Will open a new PR adding XGBoost.

**Review by marco-c — APPROVED** [review]
Thanks!
